### PR TITLE
Feature/sdk additional functions

### PR DESCRIPTION
**Pull request by Jarek-Rolski**
I added several functions to rAPId SDK and created relevant test function:

1. create_user
2. delete_user
3. list_subjects
4. list_layers
5. list_protected_domains
6. delete_dataset

I compiled and tested sdk package locally. I didn't find any issues.

### CODE REVIEW

File: sdk/rapid/rapid.py [hunk 1]
```diff
@@ -413,3 +413,130 @@ def update_subject_permissions(self, subject_id: str, permissions: list[str]):
         raise InvalidPermissionsException(
             "One or more of the provided permissions is invalid or duplicated"
         )
+
+    def create_user(self, user_name: str, user_email: str, user_permissions: list[str]):
```
Comment (MotwaniM) [inline sdk/rapid/rapid.py:417]: Looks good! The function doesn't seem to handle responses from a user issuing invalid usernames/emails with a clear error message. There are some validation functions in the API code you can steal for this

File: sdk/rapid/rapid.py [hunk 2]
```diff
@@ -413,3 +413,130 @@ def update_subject_permissions(self, subject_id: str, permissions: list[str]):
...
+            A JSON response of the API's response.
+        """
+        response = requests.get(
+            f"{self.auth.url}/protected_domains",
+            headers=self.generate_headers(),
+            timeout=TIMEOUT_PERIOD,
+        )
+        return json.loads(response.content.decode("utf-8"))
+    
+    def delete_dataset(self, layer: str, domain: str, dataset: str):
```
Comment (MotwaniM) [inline sdk/rapid/rapid.py:518]: How does this function handle incorrect permissions for deleting a dataset?

File: sdk/tests/test_rapid.py [hunk 3]
```diff
@@ -408,3 +409,95 @@ def update_subject_permissions_failure(self, requests_mock: Mocker, rapid: Rapid
...
+        res = rapid.create_user("user", "user@user.com", ["READ_ALL"])
+        assert res == mocked_response
+
+    @pytest.mark.usefixtures("requests_mock", "rapid")
+    def test_create_user_failure(self, requests_mock: Mocker, rapid: Rapid):
+        mocked_response = {"data": "dummy"}
+        requests_mock.post(f"{RAPID_URL}/user", json=mocked_response, status_code=400)
+        with pytest.raises(SubjectAlreadyExistsException):
+            rapid.create_user("user", "user@user.com", ["READ_ALL"])
+
```
Comment (MotwaniM) [inline sdk/tests/test_rapid.py:431]: Will need test for username/email being invalid.

File: sdk/tests/test_rapid.py [hunk 4]
```diff
@@ -408,3 +409,95 @@ def update_subject_permissions_failure(self, requests_mock: Mocker, rapid: Rapid
...
+    @pytest.mark.usefixtures("requests_mock", "rapid")
+    def test_list_layers(self, requests_mock: Mocker, rapid: Rapid):
+        expected = {"response": "dummy"}
+        requests_mock.get(f"{RAPID_URL}/layers", json=expected)
+
+        res = rapid.list_layers()
+        assert res == expected
+
+    @pytest.mark.usefixtures("requests_mock", "rapid")
+    def test_list_protected_domains(self, requests_mock: Mocker, rapid: Rapid):
```
Comment (MotwaniM) [inline sdk/tests/test_rapid.py:470]: Will need to test incorrect permissions for listing protected domains

File: sdk/tests/test_rapid.py [hunk 5]
```diff
@@ -408,3 +409,95 @@ def update_subject_permissions_failure(self, requests_mock: Mocker, rapid: Rapid
...
+    def test_delete_user_failure(self, requests_mock: Mocker, rapid: Rapid):
+        mocked_response = {"data": "dummy"}
+        requests_mock.delete(
+            f"{RAPID_URL}/user", json=mocked_response, status_code=400
+        )
+        with pytest.raises(SubjectNotFoundException):
+            rapid.delete_user("user", "xxx-yyy-zzz")
+
+    @pytest.mark.usefixtures("requests_mock", "rapid")
+    def test_list_subjects(self, requests_mock: Mocker, rapid: Rapid):
```
Comment (MotwaniM) [inline sdk/tests/test_rapid.py:454]: Will need to test incorrect permissions for listing subjects

File: sdk/tests/test_rapid.py [hunk 6]
```diff
@@ -408,3 +409,95 @@ def update_subject_permissions_failure(self, requests_mock: Mocker, rapid: Rapid
...
+        requests_mock.delete(
+            f"{RAPID_URL}/datasets/{layer}/{domain}/{dataset}",
+              json=mocked_response, 
+              status_code=202
+        )
+        res = rapid.delete_dataset(layer, domain, dataset)
+        assert res == mocked_response
+
+    @pytest.mark.usefixtures("requests_mock", "rapid")
+    def test_delete_dataset_failure(self, requests_mock: Mocker, rapid: Rapid):
```
Comment (MotwaniM) [inline sdk/tests/test_rapid.py:492]: Will need to test incorrect permissions for deleting dataset if user doesn't have write permissions.

**Review by MotwaniM — CHANGES_REQUESTED** [review]
Hey, looks pretty good, thank you for contributing! I've made some comments, happy to chat through them on a call.

File: sdk/rapid/rapid.py (hunk 1 above)
Comment (Jarek-Rolski) [reply sdk/rapid/rapid.py:417]: There is an issue with API status code response. For invalid permissions or username/email it returns same code.
I can cover it with if/else statement to return meaningful error anyway. 
I will also add to handle responses when user doesn't have user_admin permissions.

File: sdk/rapid/rapid.py (hunk 2 above)
Comment (Jarek-Rolski) [reply sdk/rapid/rapid.py:518]: I will cover it.
However, there is an issue with API, it returns 202 when dataset doesn't exists but layer is correct. It only returns 400 when layer is incorrect.

File: sdk/rapid/rapid.py (hunk 1 above)
Comment (Jarek-Rolski) [reply sdk/rapid/rapid.py:447]: Fixed.

File: sdk/rapid/rapid.py (hunk 2 above)
Comment (Jarek-Rolski) [reply sdk/rapid/rapid.py:582]: Fixed.

File: sdk/tests/test_rapid.py (hunk 4 above)
Comment (Jarek-Rolski) [reply sdk/tests/test_rapid.py:554]: Test added.

File: sdk/tests/test_rapid.py (hunk 6 above)
Comment (Jarek-Rolski) [reply sdk/tests/test_rapid.py]: Test added.

File: sdk/tests/test_rapid.py (hunk 5 above)
Comment (Jarek-Rolski) [reply sdk/tests/test_rapid.py]: Test added.

File: sdk/tests/test_rapid.py (hunk 3 above)
Comment (Jarek-Rolski) [reply sdk/tests/test_rapid.py:499]: Test added.

**Comment by Jarek-Rolski** [comment]
Now functions I added handle all error I was able to notice. However, there are some issues with errors/codes returned by API. 
in remove_dataset - function returns status code 202 when dataset doesn't exists, but layer is correct
in create_user - function returns status code 400 for two types of error

**Review by MotwaniM — APPROVED** [review]

**Pull request merged**